superset/utils/async_query_manager.py

```diff
@@ -0,0 +1,200 @@
+
+
+def increment_id(redis_id: str) -> str:
+    # redis stream IDs are in this format: '1607477697866-0'
+    try:
+        prefix, last = redis_id[:-1], int(redis_id[-1])
+        return prefix + str(last + 1)
+    except Exception:  # pylint: disable=broad-except
+        return redis_id
+
+
+class AsyncQueryManager:
+    MAX_EVENT_COUNT = 100
+    STATUS_PENDING = "pending"
+    STATUS_RUNNING = "running"
+    STATUS_ERROR = "error"
+    STATUS_DONE = "done"
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._redis: redis.Redis
+        self._stream_prefix: str = ""
+        self._stream_limit: Optional[int]
+        self._stream_limit_firehose: Optional[int]
+        self._jwt_cookie_name: str
+        self._jwt_cookie_secure: bool = False
+        self._jwt_secret: str
+
+    def init_app(self, app: Flask) -> None:
+        config = app.config
+        if (
+            config["CACHE_CONFIG"]["CACHE_TYPE"] == "null"
+            or config["DATA_CACHE_CONFIG"]["CACHE_TYPE"] == "null"
+        ):
+            raise Exception(
+                """
+                Cache backends (CACHE_CONFIG, DATA_CACHE_CONFIG) must be configured
+                and non-null in order to enable async queries
+                """
+            )
+
+        if len(config["GLOBAL_ASYNC_QUERIES_JWT_SECRET"]) < 32:
+            raise AsyncQueryTokenException(
+                "Please provide a JWT secret at least 32 bytes long"
+            )
+
+        self._redis = redis.Redis(  # type: ignore
+            **config["GLOBAL_ASYNC_QUERIES_REDIS_CONFIG"], decode_responses=True
+        )
+        self._stream_prefix = config["GLOBAL_ASYNC_QUERIES_REDIS_STREAM_PREFIX"]
+        self._stream_limit = config["GLOBAL_ASYNC_QUERIES_REDIS_STREAM_LIMIT"]
+        self._stream_limit_firehose = config[
+            "GLOBAL_ASYNC_QUERIES_REDIS_STREAM_LIMIT_FIREHOSE"
+        ]
+        self._jwt_cookie_name = config["GLOBAL_ASYNC_QUERIES_JWT_COOKIE_NAME"]
+        self._jwt_cookie_secure = config["GLOBAL_ASYNC_QUERIES_JWT_COOKIE_SECURE"]
+        self._jwt_secret = config["GLOBAL_ASYNC_QUERIES_JWT_SECRET"]
+
+        @app.after_request
+        def validate_session(  # pylint: disable=unused-variable
+            response: Response,
+        ) -> Response:
+            reset_token = False
+            user_id = session["user_id"] if "user_id" in session else None
+
+            if "async_channel_id" not in session or "async_user_id" not in session:
+                reset_token = True
+            elif user_id != session["async_user_id"]:
+                reset_token = True
+
+            if reset_token:
+                async_channel_id = str(uuid.uuid4())
+                session["async_channel_id"] = async_channel_id
+                session["async_user_id"] = user_id
+
+                token = self.generate_jwt(
+                    {"channel": async_channel_id, "user_id": user_id}
+                )
+
+                response.set_cookie(
```

Comment:
Can we just use FLASK_JWT_EXTENDED here? (https://flask-jwt-extended.readthedocs.io/en/stable/installation/) All of this functionality (cookie/header handling) is handled by them, and as far as I know FAB is adding it soon.

Reply:
My initial intent was to use `FLASK_JWT_EXTENDED` for this, but it was incompatible for reasons I don't now remember.

Reply:
Can we at least provide a means by which we can extend/override this functionality that doesn't involve monkey patching? I'm thinking something like defining a "JWTManager" in superset which just does these things, and then instantiate it from a string inside the `init_app` method, which would allow folks to tune this behavior. Said manager would define a few hooks like "load_jwt" "store_jwt" etc. Also, it would deal with the basic skeleton of the JWT structure, which would be customizable this way

Reply:
This JWT is narrowly-scoped in terms of functionality, and is intentionally separate from any (optional) application auth JWT usage. The main (currently only) use case here is to securely pass a channel ID between the client, Flask app and (future) websocket server. What kind of customization to you foresee being required here?

Looking back at `FLASK_JWT_EXTENDED`, I think the main issue here was that it doesn't support multiple tokens/cookies in the app, if JWT is being used as the main auth mechanism.

Reply:
I wonder have you looked into [Authlib](https://docs.authlib.org/en/latest/jose/jwt.html) before? It seems to have terrific documentation and could be incorporated for future API oauth work, too.

Reply:
Yep, we've evaluate Authlib at Preset in other contexts, but it feels like overkill for this use case.